Fix dereferencing of access type value returned from function call.
Patch by Brian Drummond.
See also https://gna.org/bugs/?18351

Index: ghdl-0.30~svn20130213/ghdl/sem_names.adb
===================================================================
--- ghdl-0.30~svn20130213.orig/ghdl/sem_names.adb	2010-01-12 04:15:20.000000000 +0100
+++ ghdl-0.30~svn20130213/ghdl/sem_names.adb	2013-12-10 22:20:28.626533000 +0100
@@ -2049,6 +2049,7 @@
    procedure Sem_Selected_By_All_Name (Name : Iir_Selected_By_All_Name)
    is
       Prefix : Iir;
+      Prefix_Name : Iir;
       Res : Iir;
 
       procedure Sem_As_Selected_By_All_Name (Sub_Name : Iir)
@@ -2071,6 +2072,7 @@
    begin
       Prefix := Get_Prefix (Name);
       Sem_Name (Prefix, True);
+      Prefix_Name := Prefix;
       Prefix := Get_Named_Entity (Prefix);
       if Prefix = Null_Iir then
          return;
@@ -2080,7 +2082,18 @@
          when Iir_Kinds_Object_Declaration
            | Iir_Kind_Selected_Element
            | Iir_Kind_Dereference
-           | Iir_Kind_Indexed_Name =>
+           | Iir_Kind_Indexed_Name
+            -- Iir_Kind_Function_Call added to resolve testcase 2 in
+            -- https://gna.org/bugs/?18351
+           | Iir_Kind_Function_Call =>
+            Sem_As_Selected_By_All_Name (Prefix);
+            -- when clause added to resolve testcases 3-6 in
+            -- https://gna.org/bugs/?18351
+         when Iir_Kinds_Function_Declaration =>
+            -- or Iir_Kind_Function_Declaration to exclude implicit functions
+            Prefix := Sem_As_Function_Call (Name => Prefix_Name,
+                                            Spec => Prefix,
+                                            Assoc_Chain => Null_Iir);
             Sem_As_Selected_By_All_Name (Prefix);
          when Iir_Kind_Error =>
             Set_Named_Entity (Name, Error_Mark);
